qiskit/pulse/logical_elements_frames.py

```diff
@@ -0,0 +1,499 @@
+    Every played pulse is associated with a ``LogicalElement`` on which it is being played.
+    Logical elements identified by their index, and a unique name for each class such that the
+    objects name is given by ``<class name><index>``.
+
+    To implement a new logical element inherit from :class:`LogicalElement` the ``name`` method needs to
+    be overridden with a proper name for the class.
+    """
+
+    def __init__(self, index):
+        """Create ``LogicalElement``.
+
+        Args:
+            index: The index of the logical element.
+        """
+        self._validate_index(index)
+        self._index = index
+        self._hash = hash(self.name)
+
+    @property
+    def index(self):
+        """Return the ``index`` of this logical element."""
+        return self._index
+
+    @abstractmethod
+    def _validate_index(self, index) -> None:
+        """Raise a PulseError if the logical element ``index`` is invalid.
+
+        Raises:
+            PulseError: If ``index`` is not valid.
+        """
+        pass
+
+    @property
+    @abstractmethod
+    def name(self) -> str:
+        """Return the name of this logical element."""
+        pass
+
+    def __repr__(self) -> str:
+        return self.name
+
+    def __eq__(self, other: "LogicalElement") -> bool:
+        """Return True iff self and other are equal, specifically, iff they have the same type
+        and the same ``identifier``.
+
+        Args:
+            other: The logical element to compare to this one.
+
+        Returns:
+            True iff equal.
+        """
+        return type(self) is type(other) and self._index == other._index
+
+    def __hash__(self) -> int:
+        return self._hash
+
+
+class Qubit(LogicalElement):
+    """Qubit logical element.
+
+    ``Qubit`` represents the different qubits in the system, as identified by
+    their (positive integer) index.
+    """
+
+    # pylint: disable=useless-parent-delegation
+    def __init__(self, index: int):
+        """Qubit logical element.
+
+        Args:
+            index: Qubit index.
+        """
+        super().__init__(index)
+
+    def _validate_index(self, index) -> None:
+        """Raise a ``PulseError`` if the qubit index is invalid. Namely, check if the index is a
+        non-negative integer.
+
+        Raises:
+            PulseError: If ``index`` is a negative integer.
+        """
+        if not isinstance(index, (int, np.integer)) or index < 0:
+            raise PulseError("Qubit index must be a non-negative integer")
+
+    @property
+    def name(self) -> str:
+        """Return the name of this qubit"""
+        return f"Q{self.index}"
+
+
+class Coupler(LogicalElement):
+    """Coupler logical element.
+
+    ``Coupler`` represents an element which couples two qubits, and can be controlled on its own.
+    It is identified by the tuple of indices of the coupled qubits.
+    """
+
+    def __init__(self, qubit_index_1: int, qubit_index_2: int):
```

Comment:
```suggestion
    def __init__(self, *qubits):
```
Coupler can take more than two qubits in principle.